File: apps/ios/Sources/DetroitQuery/Schedule.swift
// Occurrences and open-now (schema/query-spec.md "Occurrences", "Open now"). Port of packages/query/src/schedule.ts.
// The web uses the `rrule` library; here the few rule shapes the directory uses are evaluated directly, day by day,
// over a short window. Both are held to the same fixtures.
import Foundation

let closesSoonMinutes = 30
let lookaheadDays = 120

struct DayRule {
    let nth: Int?     // 2 = second, -1 = last; nil = every
    let weekday: Int  // 0 = Monday
}
private let codes = ["MO": 0, "TU": 1, "WE": 2, "TH": 3, "FR": 4, "SA": 5, "SU": 6]

func parseByday(_ s: String) -> [DayRule]? {
    var out: [DayRule] = []
    for raw in s.split(separator: ",") {
        let t = raw.trimmingCharacters(in: .whitespaces).uppercased()
        guard t.count >= 2, let wd = codes[String(t.suffix(2))] else { return nil }
        let n = String(t.dropLast(2))
        if n.isEmpty { out.append(DayRule(nth: nil, weekday: wd)) }
        else if n.range(of: #"^[+-]?[1-5]$"#, options: .regularExpression) != nil, let k = Int(n) { out.append(DayRule(nth: k, weekday: wd)) }
        else { return nil }
    }
    return out
}

func parseBymonthday(_ s: String) -> [Int]? {
    var out: [Int] = []
    for raw in s.split(separator: ",", omittingEmptySubsequences: false) {
        let t = raw.trimmingCharacters(in: .whitespaces)
        guard t.range(of: #"^-?\d{1,2}$"#, options: .regularExpression) != nil, let n = Int(t), n != 0, (-31...31).contains(n) else { return nil }
        out.append(n)
    }
    return out
}

/// The pipeline's validator (assertScheduleValid). A schedule that fails it is skipped, never guessed at.
public func scheduleIsValid(_ s: Schedule) -> Bool {
    guard parseScheduleDay(s.dtstart) != nil, parseClock(s.opensAt) != nil, parseClock(s.closesAt) != nil else { return false }
    for d in [s.until, s.validFrom, s.validTo] { if let d, parseScheduleDay(d) == nil { return false } }
    if let f = s.freq, !["DAILY", "WEEKLY", "MONTHLY", "YEARLY"].contains(f) { return false }
    if let i = s.interval, i < 1 { return false }
    // Only the shapes the spec defines, so web and iPhone can't disagree.
    if let b = s.byday {
        guard s.freq == "WEEKLY" || s.freq == "MONTHLY", let rules = parseByday(b) else { return false }
        if s.freq == "WEEKLY", rules.contains(where: { $0.nth != nil }) { return false }   // "2nd Tuesday" needs a month
    }
    if let md = s.bymonthday { guard s.freq == "MONTHLY", parseBymonthday(md) != nil else { return false } }
    return true
}

private func daysInMonth(_ y: Int, _ m: Int) -> Int { dayNumber(m == 12 ? y + 1 : y, m == 12 ? 1 : m + 1, 1) - dayNumber(y, m, 1) }

/// Does this schedule open on `day` (days since 1970-01-01)?
func opensOn(_ s: Schedule, _ day: Int) -> Bool {
    guard let start = parseDay(s.dtstart), day >= start else { return false }
    if let u = s.until.flatMap(parseDay), day > u { return false }
    guard let freq = s.freq else { return day == start }
    let every = max(1, s.interval ?? 1)
    let rules = s.byday.flatMap(parseByday)
    let c = civil(day), c0 = civil(start)
    switch freq {
    case "DAILY":
        return (day - start) % every == 0
    case "WEEKLY":
        // Weeks start on Monday (RFC 5545 default WKST); "every 2 weeks" counts from the week of dtstart.
        let week = (day - weekday(day) - (start - weekday(start))) / 7
        guard week % every == 0 else { return false }
        return rules.map { $0.contains { $0.weekday == weekday(day) } } ?? (weekday(day) == weekday(start))
    case "MONTHLY":
        guard ((c.y - c0.y) * 12 + (c.m - c0.m)) % every == 0 else { return false }
        let dim = daysInMonth(c.y, c.m)
        if let md = s.bymonthday {
            return (parseBymonthday(md) ?? []).contains { $0 > 0 ? $0 == c.d : dim + $0 + 1 == c.d }
        }
        if let rules {
            return rules.contains { r in
                guard r.weekday == weekday(day) else { return false }
                guard let n = r.nth else { return true }
                return n > 0 ? (c.d - 1) / 7 + 1 == n : (dim - c.d) / 7 + 1 == -n
            }
        }
        return c.d == c0.d
    case "YEARLY":
        return (c.y - c0.y) % every == 0 && c.m == c0.m && c.d == c0.d
    default:
        return false
    }
}

func occurrences(_ row: BundleRow, nowMin: Int, days: Int) -> [Occurrence] {
    let today = Int((Double(nowMin) / 1440).rounded(.down))
    // A row a steward marked `open_holidays` is computed as though no day were a holiday (the owner's page says so).
    let holidays = !row.flags.contains(openHolidaysFlag)
    var out: [Occurrence] = []
    for s in row.schedules where scheduleIsValid(s) {
        guard let o = parseClock(s.opensAt), let c = parseClock(s.closesAt) else { continue }
        var lo = today - 1, hi = today + days            // yesterday too, for windows that run past midnight
        if let v = s.validFrom.flatMap(parseDay) { lo = max(lo, v) }
        if let v = s.validTo.flatMap(parseDay) { hi = min(hi, v) }
        guard lo <= hi else { continue }
        for d in lo...hi where opensOn(s, d) {
            let start = d * 1440 + o
            var end = d * 1440 + c
            if end <= start { end += 1440 }
            // The occurrence belongs to the date it opens, so that is the date the holiday rule asks about: a
            // window that opened on Christmas Eve and runs to 2am is an ordinary evening's window.
            out.append(Occurrence(date: dayString(d), opensAt: s.opensAt, closesAt: s.closesAt, start: start, end: end,
                                  holiday: holidays && isHoliday(day: d)))
        }
    }
    return out.sorted { $0.start != $1.start ? $0.start < $1.start : $0.end < $1.end }
}

/// Wall-clock windows of published cancellation alerts that name this row.
func cancelWindows(_ rowId: String, _ alerts: [Alert]) -> [(start: Int, end: Int)] {
    alerts.filter { $0.kind == "cancellation" && $0.status == "published" && ($0.targets ?? []).contains(rowId) }
        .compactMap { a in
            guard let s = parseInstant(a.startsAt), let e = parseInstant(a.endsAt) else { return nil }
            return (nowWallMinutes(s), nowWallMinutes(e))
        }
}
/// An occurrence is cancelled when it overlaps a cancellation window at all, so a cancellation posted after a pantry
/// opened closes it for the rest of that window.
func isCancelled(_ o: Occurrence, _ w: [(start: Int, end: Int)]) -> Bool { w.contains { o.start < $0.end && o.end > $0.start } }

public func nextOccurrences(_ row: BundleRow, now: Date, n: Int, alerts: [Alert] = []) -> [Occurrence] {
    guard row.status == "active", row.availability == "scheduled" else { return [] }
    let nowMin = nowWallMinutes(now), w = cancelWindows(row.id, alerts)
    return Array(occurrences(row, nowMin: nowMin, days: lookaheadDays).filter { $0.end > nowMin && !isCancelled($0, w) }.prefix(n))
}

public func openNow(_ row: BundleRow, now: Date, alerts: [Alert] = []) -> OpenResult {
    guard row.status == "active" else { return OpenResult(state: .not_listed) }
    if row.availability == "always" {
        // Always open, unless a published cancellation covers right now.
        let nowMin = nowWallMinutes(now)
        return cancelWindows(row.id, alerts).contains { nowMin >= $0.start && nowMin < $0.end }
            ? OpenResult(state: .closed, next: .some(nil), cancelledNow: true) : OpenResult(state: .open)
    }
    if row.availability == "call_first" { return OpenResult(state: .call_first) }
    // Unknown is never rendered as open.
    // So is a schedule that can't be read: it is skipped, and with none left the listing is unknown.
    if row.availability == "unknown" || !row.schedules.contains(where: scheduleIsValid) { return OpenResult(state: .unknown) }

    let nowMin = nowWallMinutes(now), w = cancelWindows(row.id, alerts)
    let occ = occurrences(row, nowMin: nowMin, days: lookaheadDays).filter { $0.end > nowMin }
    var cancelledNow = false
    var current: Occurrence?
    for o in occ {
        if o.start > nowMin { break }
        if isCancelled(o, w) { cancelledNow = true; continue }
        if current == nil || o.end > current!.end { current = o }   // overlapping windows: the one open longest
    }
    if let cur = current {
        // A window that opens on a holiday tells us the usual hours and nothing about today. Never "open".
        if cur.holiday { return usualHours(cur) }
        let left = cur.end - nowMin
        return OpenResult(state: left < closesSoonMinutes ? .closes_soon : .open, closesAt: cur.closesAt, minutesLeft: left)
    }
    let next = occ.first { $0.start > nowMin && !isCancelled($0, w) }
    // "Opens later today" on a holiday is the same guess as "open now" on a holiday. A next time on a *later* day
    // stands as written: the date is a fact about the schedule, and the next-times list carries the holiday label.
    if let n = next, n.holiday, n.date == dayString(Int((Double(nowMin) / 1440).rounded(.down))) { return usualHours(n) }
    return OpenResult(state: .closed, next: .some(next.map { .init(date: $0.date, opensAt: $0.opensAt, closesAt: $0.closesAt) }), cancelledNow: cancelledNow)
}

/// The `holiday` result: the schedule's own times, offered as usual hours. No `closesAt`, nothing to count down to.
private func usualHours(_ o: Occurrence) -> OpenResult {
    OpenResult(state: .holiday, usualHours: .init(opensAt: o.opensAt, closesAt: o.closesAt))
}

File: apps/android/query/src/main/kotlin/org/help313/query/Schedule.kt
// Occurrences and open-now (schema/query-spec.md "Occurrences", "Open now").
// Port of packages/query/src/schedule.ts and apps/ios/Sources/DetroitQuery/Schedule.swift.
//
// The web app uses the `rrule` library. Here, as on iPhone, the few rule shapes the directory allows are evaluated
// directly, day by day, over a short window: 120 days of a handful of schedules is a few thousand integer
// comparisons, which is cheaper than any library and adds nothing to the APK. Both are held to the same fixtures.
package org.help313.query

private const val CLOSES_SOON_MINUTES = 30
private const val LOOKAHEAD_DAYS = 120
private const val DAY = 1440

/** nth = 2 for "second", -1 for "last", null for "every". weekday: 0 = Monday. */
internal data class DayRule(val nth: Int?, val weekday: Int)

private val CODES = mapOf("MO" to 0, "TU" to 1, "WE" to 2, "TH" to 3, "FR" to 4, "SA" to 5, "SU" to 6)
private val NTH_RE = Regex("^[+-]?[1-5]$")
private val MONTHDAY_RE = Regex("^-?\\d{1,2}$")

internal fun parseByday(s: String): List<DayRule>? {
    val out = ArrayList<DayRule>()
    for (raw in s.split(",")) {
        val t = raw.trim().uppercase()
        if (t.length < 2) return null
        val wd = CODES[t.substring(t.length - 2)] ?: return null
        val n = t.substring(0, t.length - 2)
        when {
            n.isEmpty() -> out.add(DayRule(null, wd))
            NTH_RE.matches(n) -> out.add(DayRule(n.toIntOrNull() ?: return null, wd))
            else -> return null
        }
    }
    return if (out.isEmpty()) null else out
}

internal fun parseBymonthday(s: String): List<Int>? {
    val out = ArrayList<Int>()
    for (raw in s.split(",")) {
        val t = raw.trim()
        if (!MONTHDAY_RE.matches(t)) return null
        val n = t.toIntOrNull() ?: return null
        if (n == 0 || n < -31 || n > 31) return null
        out.add(n)
    }
    return if (out.isEmpty()) null else out
}

/**
 * The pipeline's validator (assertScheduleValid). A schedule that fails it is skipped, never guessed at:
 * a guess could say "open".
 */
fun scheduleIsValid(s: Schedule): Boolean {
    if (parseScheduleDay(s.dtstart) == null) return false
    if (parseClock(s.opensAt) == null || parseClock(s.closesAt) == null) return false
    for (d in listOf(s.until, s.validFrom, s.validTo)) if (d != null && parseScheduleDay(d) == null) return false
    val freq = s.freq
    if (freq != null && freq !in setOf("DAILY", "WEEKLY", "MONTHLY", "YEARLY")) return false
    val interval = s.interval
    if (interval != null && interval < 1) return false
    // Only the shapes the spec defines, so the web, the iPhone and Android cannot disagree.
    if (s.byday != null) {
        if (freq != "WEEKLY" && freq != "MONTHLY") return false
        val rules = parseByday(s.byday) ?: return false
        // "2nd Tuesday" only means something within a month.
        if (freq == "WEEKLY" && rules.any { it.nth != null }) return false
    }
    if (s.bymonthday != null) {
        if (freq != "MONTHLY") return false
        if (parseBymonthday(s.bymonthday) == null) return false
    }
    return true
}

private fun daysInMonth(y: Int, m: Int): Int =
    dayNumber(if (m == 12) y + 1 else y, if (m == 12) 1 else m + 1, 1) - dayNumber(y, m, 1)

/** Does this schedule open on `day` (days since 1970-01-01)? */
internal fun opensOn(s: Schedule, day: Int): Boolean {
    val start = parseDay(s.dtstart) ?: return false
    if (day < start) return false
    val until = s.until?.let { parseDay(it) }
    if (until != null && day > until) return false
    val freq = s.freq ?: return day == start
    val every = maxOf(1, s.interval ?: 1)
    val rules = s.byday?.let { parseByday(it) }
    val c = civil(day)
    val c0 = civil(start)
    return when (freq) {
        "DAILY" -> (day - start) % every == 0
        "WEEKLY" -> {
            // Weeks start on Monday (RFC 5545 default WKST); "every 2 weeks" counts from the week of dtstart.
            val week = (day - weekday(day) - (start - weekday(start))) / 7
            if (week % every != 0) false
            else if (rules != null) rules.any { it.weekday == weekday(day) }
            else weekday(day) == weekday(start)
        }
        "MONTHLY" -> {
            if (((c.y - c0.y) * 12 + (c.m - c0.m)) % every != 0) return false
            val dim = daysInMonth(c.y, c.m)
            val md = s.bymonthday
            when {
                md != null -> (parseBymonthday(md) ?: emptyList()).any { if (it > 0) it == c.d else dim + it + 1 == c.d }
                rules != null -> rules.any { r ->
                    if (r.weekday != weekday(day)) false
                    else when (val n = r.nth) {
                        null -> true
                        else -> if (n > 0) (c.d - 1) / 7 + 1 == n else (dim - c.d) / 7 + 1 == -n
                    }
                }
                else -> c.d == c0.d
            }
        }
        "YEARLY" -> (c.y - c0.y) % every == 0 && c.m == c0.m && c.d == c0.d
        else -> false
    }
}

internal fun occurrences(row: BundleRow, nowMin: WallMinutes, days: Int): List<Occurrence> {
    val today = Math.floorDiv(nowMin, DAY)
    val out = ArrayList<Occurrence>()
    for (s in row.schedules) {
        if (!scheduleIsValid(s)) continue
        val o = parseClock(s.opensAt) ?: continue
        val c = parseClock(s.closesAt) ?: continue
        var lo = today - 1              // yesterday too, for windows that run past midnight
        var hi = today + days
        s.validFrom?.let { parseDay(it) }?.let { if (it > lo) lo = it }
        s.validTo?.let { parseDay(it) }?.let { if (it < hi) hi = it }
        if (lo > hi) continue
        for (d in lo..hi) {
            if (!opensOn(s, d)) continue
            val start = d * DAY + o
            var end = d * DAY + c
            if (end <= start) end += DAY
            out.add(Occurrence(dayString(d), s.opensAt, s.closesAt, start, end))
        }
    }
    out.sortWith(compareBy({ it.start }, { it.end }))
    return out
}

internal data class CancelWindow(val start: WallMinutes, val end: WallMinutes)

/** Wall-clock windows of published cancellation alerts that name this row. */
internal fun cancelWindows(rowId: String, alerts: List<Alert>): List<CancelWindow> =
    alerts.filter { it.kind == "cancellation" && it.status == "published" && it.targets.contains(rowId) }
        .mapNotNull {
            val s = parseInstant(it.startsAt) ?: return@mapNotNull null
            val e = parseInstant(it.endsAt) ?: return@mapNotNull null
            CancelWindow(nowWallMinutes(s), nowWallMinutes(e))
        }

/**
 * An occurrence is cancelled when it overlaps a cancellation window at all, so a cancellation posted after a
 * pantry opened closes it for the rest of that window.
 */
internal fun isCancelled(o: Occurrence, windows: List<CancelWindow>): Boolean =
    windows.any { o.start < it.end && o.end > it.start }

fun nextOccurrences(row: BundleRow, nowMillis: Long, n: Int, alerts: List<Alert> = emptyList()): List<Occurrence> {
    if (row.status != "active" || row.availability != "scheduled") return emptyList()
    val nowMin = nowWallMinutes(nowMillis)
    val windows = cancelWindows(row.id, alerts)
    return occurrences(row, nowMin, LOOKAHEAD_DAYS)
        .filter { it.end > nowMin && !isCancelled(it, windows) }
        .take(n)
}

fun openNow(row: BundleRow, nowMillis: Long, alerts: List<Alert> = emptyList()): OpenResult {
    if (row.status != "active") return OpenResult(OpenState.NOT_LISTED)
    if (row.availability == "always") {
        // Always open, unless a published cancellation covers right now.
        val nowMin = nowWallMinutes(nowMillis)
        val covered = cancelWindows(row.id, alerts).any { nowMin >= it.start && nowMin < it.end }
        return if (covered) OpenResult(OpenState.CLOSED, next = null, cancelledNow = true) else OpenResult(OpenState.OPEN)
    }
    if (row.availability == "call_first") return OpenResult(OpenState.CALL_FIRST)
    // Unknown is never rendered as open. Nor is a schedule that cannot be read: it is skipped, and with none
    // left the listing is unknown.
    if (row.availability == "unknown" || row.schedules.none { scheduleIsValid(it) }) return OpenResult(OpenState.UNKNOWN)

    val nowMin = nowWallMinutes(nowMillis)
    val windows = cancelWindows(row.id, alerts)
    val occ = occurrences(row, nowMin, LOOKAHEAD_DAYS).filter { it.end > nowMin }

    var cancelledNow = false
    var current: Occurrence? = null
    for (o in occ) {
        if (o.start > nowMin) break
        if (isCancelled(o, windows)) { cancelledNow = true; continue }
        // Overlapping windows: keep the one that stays open longest.
        val held = current
        if (held == null || o.end > held.end) current = o
    }
    val cur = current
    if (cur != null) {
        val left = cur.end - nowMin
        return OpenResult(
            state = if (left < CLOSES_SOON_MINUTES) OpenState.CLOSES_SOON else OpenState.OPEN,
            closesAt = cur.closesAt,
            minutesLeft = left,
        )
    }
    val next = occ.firstOrNull { it.start > nowMin && !isCancelled(it, windows) }
    return OpenResult(
        state = OpenState.CLOSED,
        next = next?.let { NextTime(it.date, it.opensAt, it.closesAt) },
        cancelledNow = cancelledNow,
    )
}
